src/controllers/aer_controller.hpp

```diff
@@ -1404,85 +1525,262 @@ Transpile::Fusion Controller::transpile_fusion(Method method,
-  state.set_parallalization(parallel_state_update_);
-  state.set_distribution(num_processes_);
-  state.set_global_phase(circ.global_phase_angle);
+  // Execute in try block so we can catch errors and return the error message
+  // for individual circuit failures.
+  try {
+    // Rng engine (this one is used to add noise on circuit)
+    RngEngine rng;
+    rng.set_seed(rng_seed);
+
+    // Output data container
+    result.set_config(config);
+    result.metadata.add(method_names_.at(method), "method");
+    if (method == Method::statevector || method == Method::density_matrix ||
+        method == Method::unitary) {
+      result.metadata.add(sim_device_name_, "device");
+    } else {
+      result.metadata.add("CPU", "device");
+    }
 
-  // Rng engine
-  RngEngine rng;
-  rng.set_seed(rng_seed);
+    // Circuit qubit metadata
+    result.metadata.add(circ.num_qubits, "num_qubits");
+    result.metadata.add(circ.num_memory, "num_clbits");
+    result.metadata.add(circ.qubits(), "active_input_qubits");
+    result.metadata.add(circ.qubit_map(), "input_qubit_map");
+    result.metadata.add(circ.remapped_qubits, "remapped_qubits");
+
+    // Add measure sampling to metadata
+    // Note: this will set to `true` if sampling is enabled for the circuit
+    result.metadata.add(false, "measure_sampling");
+    result.metadata.add(false, "batched_shots_optimization");
+
+    if(circ.num_qubits > 0){  //do nothing for query steps
+      // Choose execution method based on noise and method
+      Circuit opt_circ;
+
+      bool noise_sampling = false;
+      // Ideal circuit
+      if (noise.is_ideal()) {
+        opt_circ = circ;
+        result.metadata.add("ideal", "noise");
+      }
+      // Readout error only
+      else if (noise.has_quantum_errors() == false) {
+        opt_circ = noise.sample_noise(circ, rng);
+        result.metadata.add("readout", "noise");
+      }
+      // Superop noise sampling
+      else if (method == Method::density_matrix || method == Method::superop) {
+        // Sample noise using SuperOp method
+        opt_circ = noise.sample_noise(circ, rng, Noise::NoiseModel::Method::superop);
+        result.metadata.add("superop", "noise");
+      }
+      // Kraus noise sampling
+      else if (noise.opset().contains(Operations::OpType::kraus) ||
+               noise.opset().contains(Operations::OpType::superop)) {
+        opt_circ = noise.sample_noise(circ, rng, Noise::NoiseModel::Method::kraus);
+        result.metadata.add("kraus", "noise");
+      }
+      // General circuit noise sampling
+      else {
+        if(batched_shots_optimization_ && batched_shots_optimization_threshold_ >= max_qubits_ && 
+           sim_device_ == Device::GPU && max_batched_states_ > 1 && max_batched_states_ >= num_gpus_){
+          //for GPU noise sampling is done at runtime
+          opt_circ = noise.sample_noise(circ, rng, Noise::NoiseModel::Method::circuit, true);
+          opt_circ.can_sample = false;
```

Comment:
`opt_circ.can_sample = false` should be set in `noise.sample_noise()`.

Reply:
removed from controller